File: source/clojure/src/tarjan.clj
(ns tarjan)

(set! *warn-on-reflection* true)
(set! *unchecked-math* :warn-on-boxed)

(def algorithm-name "Tarjan")

(defn create
  ^ints [^long n]
  (let [root (int-array n)]
    (dotimes [i n]
      (aset-int root i i))
    root))

(defn find-root
  ^long [^ints root ^long a ^ints path-scratch]
  (loop [current a
         path-len 0]
    (let [parent (aget root current)]
      (if (= parent current)
        (do
          (dotimes [i path-len]
            (aset-int root (aget path-scratch i) current))
          current)
        (do
          (aset-int path-scratch path-len current)
          (recur parent (inc path-len)))))))

(defn unite!
  [^objects tarjan-state ^long a ^long b]
  (let [^ints root (aget tarjan-state 0)
        ^longs sets-ref (aget tarjan-state 1)
        ^ints path-scratch (aget tarjan-state 2)
        a-root (find-root root a path-scratch)
        b-root (find-root root b path-scratch)]
    (when (not= a-root b-root)
      (aset-int root a-root b-root)
      (aset sets-ref 0 (dec (aget sets-ref 0))))))

(defn reset-state!
  [^ints root ^longs sets-ref ^long n]
  (dotimes [i n]
    (aset-int root i i))
  (aset sets-ref 0 n))

(defn set-count
  ^long [^longs sets-ref]
  (aget sets-ref 0))

File: source/nim/src/tarjan.nim
import std/sequtils

type
  Tarjan* = object
    root: seq[int]
    sets: int

proc create*(n: int): Tarjan =
  result.root = toSeq(0..<n)
  result.sets = n

proc reset*(rel: var Tarjan, n: int) =
  for i in 0..<n:
    rel.root[i] = i
  rel.sets = n

proc find(rel: var Tarjan, a: int): int =
  var current = a
  
  while rel.root[current] != current:
    current = rel.root[current]
  
  result = current
  current = a
  
  while rel.root[current] != result:
    let next = rel.root[current]
    rel.root[current] = result
    current = next

proc unite*(rel: var Tarjan, a, b: int) =
  let a = rel.find(a)
  let b = rel.find(b)
  
  if a != b:
    dec rel.sets
    rel.root[a] = b

proc setCount*(rel: Tarjan): int =
  rel.sets

const name* = "Tarjan"
